cluster-controller/src/main/java/io/strimzi/controller/cluster/resources/TopicController.java

```diff
@@ -0,0 +1,245 @@
+        this.config
+                .withImage(DEFAULT_IMAGE)
+                .withKafkaBootstrapServers(defaultBootstrapServers(cluster))
+                .withZookeeperConnect(defaultZookeeperConnect(cluster))
+                .withNamespace(namespace)
+                .withReconciliationInterval(DEFAULT_FULL_RECONCILIATION_INTERVAL)
+                .withZookeeperSessionTimeout(DEFAULT_ZOOKEEPER_SESSION_TIMEOUT);
+    }
+
+    public static String topicControllerName(String cluster) {
+        return cluster + TopicController.NAME_SUFFIX;
+    }
+
+    protected static String defaultZookeeperConnect(String cluster) {
+        return cluster + ZookeeperCluster.NAME_SUFFIX + ":" + DEFAULT_ZOOKEEPER_PORT;
+    }
+
+    protected static String defaultBootstrapServers(String cluster) {
+        return cluster + KafkaCluster.NAME_SUFFIX + ":" + DEFAULT_BOOTSTRAP_SERVERS_PORT;
+    }
+
+    public void setConfig(TopicControllerConfig config) {
+        this.config = config;
+    }
+
+    public TopicControllerConfig getConfig() {
+        return this.config;
+    }
+
+    /**
+     * Create a Topic Controller from the related ConfigMap resource
+     *
+     * @param kafkaClusterCm ConfigMap with cluster configuration containing the topic controller one
+     * @return Topic Controller instance, null if not configured in the ConfigMap
+     */
+    public static TopicController fromConfigMap(ConfigMap kafkaClusterCm) {
+
+        TopicController topicController = null;
+
+        String config = kafkaClusterCm.getData().get(KEY_CONFIG);
+        if (config != null) {
+            topicController = new TopicController(kafkaClusterCm.getMetadata().getNamespace(), kafkaClusterCm.getMetadata().getName());
+            topicController.setLabels(kafkaClusterCm.getMetadata().getLabels());
+
+            TopicControllerConfig.fromJson(topicController.getConfig(), new JsonObject(config));
+            topicController.setImage(topicController.getConfig().image());
+        }
+
+        return topicController;
+    }
+
+    /**
+     * Create a Topic Controller from the deployed Deployment resource
+     *
+     * @param namespace Kubernetes/OpenShift namespace where cluster resources are going to be created
+     * @param cluster overall cluster name
+     * @param dep the deployment from which to recover the topic controller state
+     * @return Topic Controller instance, null if the corresponding Deployment doesn't exist
+     */
+    public static TopicController fromDeployment(String namespace, String cluster, Deployment dep) {
+
+        TopicController topicController = null;
+
+        if (dep != null) {
+
+            topicController = new TopicController(namespace, cluster);
+
+            topicController.setLabels(dep.getMetadata().getLabels());
+            topicController.setReplicas(dep.getSpec().getReplicas());
+            topicController.setImage(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getImage());
```

Comment:
Can we extract a local for the repeated `dep.getSpec().getTemplate().getSpec().getContainers().get(0)`?

Reply:
Agree. It's something that happens in other classes as well. Should it be better to do the changes in another PR all together ?

Reply:
Happy to have a follow up PR for this.